product_provider/app.py

```diff
@@ -1,42 +1,183 @@
 
-    context: object, required
-        Lambda Context runtime methods and attributes
+  properties["ProvisioningArtifactAction"] = "ALL"
+  provisioning_artifact_action = resource_properties.get("ProvisioningArtifactAction")
+  if provisioning_artifact_action:
+    if provisioning_artifact_action not in ["ALL", "ALL_EXCEPT_LATEST"] :
+      raise ValueError(f"ProvisioningArtifactAction valid values are ALL|ALL_EXCEPT_LATEST")
+    else:
+      properties["ProvisioningArtifactAction"] = provisioning_artifact_action
 
-        Context doc: https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+  return properties
 
-    Returns
-    ------
-    API Gateway Lambda Proxy Output Format: dict
+def get_latest_provisioning_artifact(provisioning_artifacts):
+  '''
+  Get the latest the latest version of a product's provisioning artifact.
+  :param provisioning_artifacts: a list from get_provisioning_artifacts
+  :return: a dict with the product's info about the latest provisioning artifact
+  '''
+  versions = []
+  for provisioning_artifact in provisioning_artifacts:
+    versions.append(provisioning_artifact['Name'])
 
-        Return doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html
-    """
+  latest_version = max(versions, key=VersionInfo.parse)
+  log.debug(f"latest version: {latest_version}")
 
-    # try:
-    #     ip = requests.get("http://checkip.amazonaws.com/")
-    # except requests.RequestException as e:
-    #     # Send some context about this error to Lambda Logs
-    #     print(e)
+  for provisioning_artifact in provisioning_artifacts:
+    if VersionInfo.parse(latest_version) == VersionInfo.parse(provisioning_artifact['Name']):
+      return provisioning_artifact
 
-    #     raise e
 
-    return {
-        "statusCode": 200,
-        "body": json.dumps({
-            "message": "hello world",
-            # "location": ip.text.replace("\n", "")
-        }),
+def get_provisioning_artifacts(product_info):
+  '''
+  Takes a list of SC product info and generates a list of provisioning
+  artifacts with semantic version numbers in the Name field.
+
+  We assume that the product's provisioning artifact summary names contains
+  a version number, something (i.e. v1.0.0, v1.0.2, etc..).  We strip the `v`
+  from the version names to conform to semantic versions so we can use it
+  to compare version numbers.
+
+  :param product_info: response from https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/servicecatalog.html#ServiceCatalog.Client.describe_product_as_admin
+  :return: A list of the product's provisioned artifacts
+  '''
+  ProductId = product_info['ProductViewDetail']['ProductViewSummary']['ProductId']
+  summaries = product_info['ProvisioningArtifactSummaries']
+  artifacts = []
+  for summary in summaries:
+    # remove `v` from the name to get semver
+    semantic_version = summary['Name'][1:]
```

Comment:
May want to check whether the string starts with v before assuming `[1:]`

Reply:
it doesn't matter what the name is, we check that it's a valid semantic version on the next line.